### PR TITLE
WIP - Feat bilateral swap

### PR DESCRIPTION
**Pull request by thomas-waite**
## Description
Integrated all the zero knowledge bilateral swap work into the monorepo. This includes Javascript proof construction and verification code, together with the smart contract verifier. There are tests for all new scripts.

## Packages Updated
**aztec.js**:
1) Created bilateralSwap (aztec.js/src/proof/bilateralSwap). This has the Javascript zk proof construction and verification code (in index.js). It also contains a helpers.js script to provide additional functionality.
Created bilateralSwapProof.spec.js (aztec.js/src/test/proof/bilateralSwapProof.spec.js). This has a suite of tests to validate the functionality of the Javascript proof construction and verification code.

**contracts**
1) Created BilateralSwap.sol (contracts/contracts/AZTEC/BilateralSwap.sol). This is a smart contract, implemented in YUL, that validates a zero knowledge bilateral swap
Created bilateralSwap.js (contracts/test/AZTEC/bilateralSwap.js). This contains a suite of tests to validate performance of the BilateralSwap.sol smart contract

# #Checklist

- [- ] Write tests for added code.
- [ -] Prefix PR title with [WIP] if necessary.
- [ -] Update documentation when needed.
- [ ] Add new entries to the relevant CHANGELOG.json files.

### CODE REVIEW

File: packages/contracts/test/AZTEC/bilateralSwap.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            const fakeProofData = new Array(4).map(() => new Array(6).map(() => toBytes32.randomBytes32()));
+
+            await exceptions.catchRevert(bilateralSwapContract.validateBilateralSwap(fakeProofData, challenge, t2, {
+                from: accounts[0],
+                gas: 4000000,
+            }));
+        });
+
+        it('Validate failure when points not on curve', async () => {
+            const zeroes = `${padLeft('0', 64)}`;
```
Comment (zac-williamson) [inline packages/contracts/test/AZTEC/bilateralSwap.js]: We should construct a proof that would pass were it not for `validateCommitment` - see comments on `bilateralSwapProof.spec.js`

File: packages/contracts/test/AZTEC/bilateralSwap.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+contract('BilateralSwap', (accounts) => {
+    describe('success states', () => {
+        let bilateralSwap;
+        let testNotes;
+
+        beforeEach(async () => {
+            bilateralSwap = await BilateralSwap.new(accounts[0]);
+            const makerNoteValues = [10, 20];
+            const takerNoteValues = [10, 20];
```
Comment (zac-williamson) [inline packages/contracts/test/AZTEC/bilateralSwap.js]: We should generate note values randomly distributed from [0, ..., K_MAX] (I need to update my tests in `aztec.js` to do this as well!)

File: packages/aztec.js/src/proof/bilateralSwap/helpers.js [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+};
+
+helpers.makeNoteArray = (notes) => {
+    const makerNotes = Object.values(notes.makerNotes);
+    const takerNotes = Object.values(notes.takerNotes);
+    const noteArray = [makerNotes[0], makerNotes[1], takerNotes[0], takerNotes[1]];
+    return noteArray;
+};
+
+helpers.makeIncorrectArray = (notes) => {
```
Comment (zac-williamson) [inline packages/aztec.js/src/proof/bilateralSwap/helpers.js]: I can't see this being used anywhere, should be deleted

File: packages/aztec.js/src/proof/bilateralSwap/helpers.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        - The purpose is to set bk1 = bk3 and bk2 = bk4
+        - i is used as an indexing variable, to keep track of whether we are at a maker note or taker note
+        - All bks are stored in a bkArray. When we arrive at the taker notes, we set bk equal to the bk of the corresponding 
+          maker note. This is achieved by 'jumping back' 2 index positions (i - 2) in the bkArray, and setting the current
+          bk equal to the element at the resulting position.
+        */
+
+        // Maker notes
+        if (i <= 1) {
+            B = note.gamma.mul(bk).add(bn128.h.mul(ba));
```
Comment (zac-williamson) [inline packages/aztec.js/src/proof/bilateralSwap/helpers.js]: Computation of `B` should be outside the `if, else` block

File: packages/aztec.js/src/proof/bilateralSwap/index.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+    finalHash.appendBN(new BN(sender.slice(2), 16));
+
+    notes.forEach((note) => {
+        finalHash.append(note.gamma);
+        finalHash.append(note.sigma);
+    });
+
+    const bkArray = [];
+    const blindingFactors = notes.map((note, i) => {
```
Comment (zac-williamson) [inline packages/aztec.js/src/proof/bilateralSwap/index.js]: This looks like the same code as `helpers.getBlindingFactorsAndChallenge`, we should re-use that method if this is the case.

File: packages/contracts/contracts/AZTEC/BilateralSwap.sol [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+             * input notes and output notes. This is used both as an input to 
+             * validate the challenge `c` and also to generate pseudorandom relationships
+             * between commitments for different outputNotes, so that we can combine 
+             * them into a single multi-exponentiation for the purposes of validating 
+             * the bilinear pairing relationships.
+             * @param notes calldata location notes
+             * @param n number of notes
+             **/
+            function hashCommitments(notes, n) {
+                for { let i := 0 } lt(i, n) { i := add(i, 0x01) } {
```
Comment (zac-williamson) [inline packages/contracts/contracts/AZTEC/BilateralSwap.sol]: We know `n=4`, so we can un-roll this loop and use hard-coded offsets to index calldata - saves ~300 gas.

```
calldatacopy(0x300, add(notes, 0x60), 0x80)
calldatacopy(0x380, add(notes, 0x120), 0x80)
calldatacopy(0x400, add(notes, 0x1e0), 0x80)
calldatacopy(0x480, add(notes, 0x2a0), 0x80)
mstore(0x00, keccak256(0x300, 0x200))
```

File: packages/aztec.js/src/proof/bilateralSwap/index.js [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        bkArray.push(bk);
+
+        return {
+            bk,
+            ba,
+            B,
+        };
+    });
+
+    const challenge = finalHash.keccak(groupReduction);
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/index.js]: Done

File: packages/aztec.js/src/proof/bilateralSwap/helpers.js [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+
+const helpers = {};
+
+helpers.makeTestNotes = (makerNoteValues, takerNoteValues) => {
+    const noteValues = makerNoteValues.concat(takerNoteValues);
+    const numNotes = noteValues.length;
+
+    let i;
+    const publicKeys = [];
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/helpers.js]: Far nicer way of doing it

File: packages/aztec.js/src/proof/bilateralSwap/helpers.js [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+const notesConstruct = require('../../note');
+
+
+const { groupReduction } = bn128;
+
+
+const helpers = {};
+
+helpers.makeTestNotes = (makerNoteValues, takerNoteValues) => {
+    const noteValues = makerNoteValues.concat(takerNoteValues);
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/helpers.js]: Didn't realise, good idea

File: packages/aztec.js/src/proof/bilateralSwap/helpers.js [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    const testNotes = publicKeys.map((publicKey, j) => {
+        return notesConstruct.create(publicKey, noteValues[j]);
+    });
+
+    return testNotes;
+};
+
+helpers.checkNumberNotes = (notes, numberExpected) => {
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/helpers.js]: Done

File: packages/aztec.js/src/proof/bilateralSwap/helpers.js [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+    if (notes.length !== numberExpected) {
+        throw new Error('Incorrect number of notes');
+    }
+};
+
+helpers.makeNoteArray = (notes) => {
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/helpers.js]: Good spot - remnant from an old way of constructing test notes

File: packages/aztec.js/src/proof/bilateralSwap/helpers.js [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+};
+
+helpers.makeIncorrectArray = (notes) => {
+    const makerNotes = Object.values(notes.makerNotes);
+    const takerNotes = Object.values(notes.takerNotes);
+    const noteArray = [makerNotes[0], makerNotes[1], makerNotes[2], takerNotes[0], takerNotes[1], takerNotes[2]];
+    return noteArray;
+};
+
+helpers.validateOnCurve = (x, y) => {
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/helpers.js]: Implemented

File: packages/aztec.js/src/proof/bilateralSwap/helpers.js [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+            bk,
+            ba,
+            B,
+        };
+    });
+    const challenge = finalHash.keccak(groupReduction);
+    return { blindingFactors, challenge };
+};
+
+helpers.toBnAndAppendPoints = (proofData) => {
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/helpers.js]: Sure, will leave on the back burner for now

File: packages/aztec.js/src/proof/bilateralSwap/helpers.js [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+        - The purpose is to set kBar1 = kBar3 and kBar2 = kBar4
+        - i is used as an indexing variable, to keep track of whether we are at a maker note or taker note
+        - All kBars are stored in a kBarArray. When we arrive at the taker notes, we set bk equal to the bk of the corresponding 
+          maker note. This is achieved by 'jumping back' 2 index positions (i - 2) in the kBarArray, and setting the current
+          kBar equal to the element at the resulting position.
+        */
+
+        // Maker notes
+        if (i <= 1) {
+            B = gamma.mul(kBar).add(bn128.h.mul(aBar)).add(sigma.mul(formattedChallenge).neg());
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/helpers.js]: Done

File: packages/aztec.js/src/proof/bilateralSwap/index.js [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+
+/**
+ * Verify AZTEC bilateral swap proof transcript
+ *
+ * @method verifyBilateralSwap
+ * @param {Array[proofData]} proofData - proofData array of AZTEC notes
+ * @param {big number instance} challenge - challenge variable used in zero-knowledge protocol 
+ * @returns {number} - returns 1 if proof is validated, throws an error if not
+ */
+bilateralSwap.verifyBilateralSwap = (proofData, challenge, sender) => {
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/index.js]: Done

File: packages/aztec.js/src/proof/bilateralSwap/index.js [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+            bk,
+            ba,
+            B,
+        };
+    });
+
+    const challenge = finalHash.keccak(groupReduction);
+
+    const proofData = blindingFactors.map((blindingFactor, i) => {
+        const kBar = ((notes[i].k.redMul(challenge)).redAdd(blindingFactor.bk)).fromRed();
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/index.js]: Done

File: packages/aztec.js/test/proof/bilateralSwapProof.spec.js [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+const { expect } = chai;
+
+
+describe('Validating bilateral swap proof construction and verification algos', () => {
+    describe('Validate properties of the proof construction algo', () => {
+        let testNotes;
+        let sender;
+
+        beforeEach(() => {
+            testNotes = helpers.makeTestNotes([10, 20], [10, 20]);
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/test/proof/bilateralSwapProof.spec.js]: Done

File: packages/contracts/contracts/AZTEC/BilateralSwap.sol [hunk 18]
```diff
@@ -0,0 +1,264 @@
+pragma solidity ^0.4.24;
+
+library BilateralSwapInterface {
+    function validateBilateralSwap(bytes32[6][], uint, bytes32[4]) external pure returns (bool) {}
+                                // 6 pieces of data per note, unspecified number of notes
+}
+
+/**npm aud
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/contracts/contracts/AZTEC/BilateralSwap.sol]: Done

File: packages/contracts/contracts/AZTEC/BilateralSwap.sol [hunk 19]
```diff
@@ -0,0 +1,264 @@
...
+
+                }
+
+                let gen_order := 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001
+                let challenge := mod(calldataload(0x24), gen_order)
+
+                mstore(0x2e0, caller) // store the msg.sender, to be hashed later
+
+                hashCommitments(notes, n)
+                let b := add(0x300, mul(n, 0x80)) // set pointer to memory location of commitments where the commitments
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/contracts/contracts/AZTEC/BilateralSwap.sol]: Done

File: packages/contracts/contracts/AZTEC/BilateralSwap.sol [hunk 20]
```diff
@@ -0,0 +1,264 @@
...
+             * @param notes calldata location notes
+             * @param n number of notes
+             **/
+            function hashCommitments(notes, n) {
+                for { let i := 0 } lt(i, n) { i := add(i, 0x01) } {
+                    let index := add(add(notes, mul(i, 0xc0)), 0x60)
+
+                    calldatacopy(add(0x300, mul(i, 0x80)), index, 0x80)
+                }
+                // storing at position 0x00 in memory, the kecca hash of everything from
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/contracts/contracts/AZTEC/BilateralSwap.sol]: Done

File: packages/contracts/test/AZTEC/bilateralSwap.js [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+Outline of test:
+- This needs to be able to run on Truffle - to test the smart contract feature
+- Construct success states
+- Generate a proof using the BilateralSwap javascript proof generation code
+- Validate that it works, first using the javascript proof verification code
+- Validate that it works, using the bilateral swap smart construct
+*/
+
+contract('BilateralSwap', (accounts) => {
+    describe('success states', () => {
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/contracts/test/AZTEC/bilateralSwap.js]: Done

File: packages/aztec.js/src/proof/bilateralSwap/helpers.js (hunk 3 above)
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/helpers.js]: Done

File: packages/aztec.js/src/proof/bilateralSwap/helpers.js (hunk 4 above)
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/helpers.js]: Done

File: packages/contracts/contracts/AZTEC/BilateralSwap.sol (hunk 6 above)
Comment (thomas-waite) [reply packages/contracts/contracts/AZTEC/BilateralSwap.sol]: Done

File: packages/aztec.js/test/proof/bilateralSwapProof.spec.js [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+
+const bilateralProof = require('../../src/proof/bilateralSwap');
+const helpers = require('../../src/proof/bilateralSwap/helpers');
+const Keccak = require('../../src/keccak');
+
+
+const { expect } = chai;
+
+
+describe('Validating bilateral swap proof construction and verification algos', () => {
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/test/proof/bilateralSwapProof.spec.js]: Done

File: packages/aztec.js/src/proof/bilateralSwap/index.js [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+
+    const kBarArray = [];
+
+    // Validate that the commitments lie on the bn128 curve
+    proofDataBn.forEach((proofElement) => {
+        helpers.validateOnCurve(proofElement[2], proofElement[3]); // checking gamma point
+        helpers.validateOnCurve(proofElement[4], proofElement[5]); // checking sigma point
+    });
+
+    proofDataBn.map((proofElement, i) => {
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/index.js]: They do slightly different things, although they are fairly similar. Inclined to leave as is

File: packages/aztec.js/src/proof/bilateralSwap/index.js (hunk 5 above)
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/index.js]: They do slightly different things, although they are similar. Inclined to leave as is

File: packages/aztec.js/test/proof/bilateralSwapProof.spec.js [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+            const { proofData } = bilateralProof.constructBilateralSwap(testNotes, sender);
+            expect(proofData.length).to.equal(4);
+            expect(proofData[0].length).to.equal(6);
+            expect(proofData[1].length).to.equal(6);
+            expect(proofData[2].length).to.equal(6);
+            expect(proofData[3].length).to.equal(6);
+        });
+
+        it('validate that the proof is correct, using the validation algo', () => {
+            const { proofData, challenge } = bilateralProof.constructBilateralSwap(testNotes, sender);
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/test/proof/bilateralSwapProof.spec.js]: Done

File: packages/contracts/test/AZTEC/bilateralSwap.js (hunk 2 above)
Comment (thomas-waite) [reply packages/contracts/test/AZTEC/bilateralSwap.js]: Done

File: packages/contracts/test/AZTEC/bilateralSwap.js (hunk 1 above)
Comment (thomas-waite) [reply packages/contracts/test/AZTEC/bilateralSwap.js]: Done

**Pull request merged**